python/sglang/srt/layers/moe/moe_runner/marlin.py

```diff
@@ -0,0 +1,170 @@
+    w2_g_idx: Optional[torch.Tensor] = None
+    is_k_full: bool = True
+
+    # AWQ specific (Optional)
+    w13_qzeros: Optional[torch.Tensor] = None
+    w2_qzeros: Optional[torch.Tensor] = None
+
+    # Optional
+    expert_map: Optional[torch.Tensor] = None
+
+
+class MarlinRunnerCore(MoeRunnerCore):
+    def __init__(self, config: MoeRunnerConfig):
+        super().__init__(config)
+        self.workspace: Optional[torch.Tensor] = None
+
+    def run(
+        self,
+        runner_input: MarlinRunnerInput,
+        quant_info: MarlinMoeQuantInfo,
+        running_state: dict,
+    ) -> MarlinRunnerOutput:
+        from sglang.srt.layers.moe.fused_moe_triton.fused_marlin_moe import (
+            fused_marlin_moe,
+        )
+        from sglang.srt.layers.quantization.marlin_utils import marlin_make_workspace
+
+        x = runner_input.hidden_states
+
+        assert self.config.activation == "silu", "Only SiLU activation is supported."
+
+        if self.workspace is None:
+            self.workspace = marlin_make_workspace(x.device, max_blocks_per_sm=4)
+
+        output = fused_marlin_moe(
+            hidden_states=x,
+            w1=quant_info.w13_qweight,
+            w2=quant_info.w2_qweight,
+            w1_scale=quant_info.w13_scales,
+            w2_scale=quant_info.w2_scales,
+            gating_output=runner_input.router_logits,
+            topk_weights=runner_input.topk_weights,
+            topk_ids=runner_input.topk_ids,
+            expert_map=quant_info.expert_map,
+            g_idx1=quant_info.w13_g_idx,
+            g_idx2=quant_info.w2_g_idx,
+            sort_indices1=quant_info.w13_g_idx_sort_indices,
+            sort_indices2=quant_info.w2_g_idx_sort_indices,
+            w1_zeros=quant_info.w13_qzeros,
+            w2_zeros=quant_info.w2_qzeros,
+            workspace=self.workspace,
+            num_bits=quant_info.weight_bits,
+            is_k_full=quant_info.is_k_full,
+            inplace=self.config.inplace,
+            routed_scaling_factor=self.config.routed_scaling_factor,
+        ).to(x.dtype)
+
+        return MarlinRunnerOutput(hidden_states=output)
+
+    @property
+    def runner_backend(self) -> MoeRunnerBackend:
+        return MoeRunnerBackend.MARLIN
+
+
+@register_fused_func("none", "marlin")
+def fused_experts_none_to_marlin(
+    dispatch_output: StandardDispatchOutput,
+    quant_info: MarlinMoeQuantInfo,
+    runner_config: MoeRunnerConfig,
+) -> StandardCombineInput:
+    from sglang.srt.layers.moe.fused_moe_triton.fused_marlin_moe import fused_marlin_moe
+    from sglang.srt.layers.moe.token_dispatcher.standard import StandardCombineInput
+    from sglang.srt.layers.quantization.marlin_utils import marlin_make_workspace
+
+    hidden_states = dispatch_output.hidden_states
+    topk_output = dispatch_output.topk_output
+
+    assert runner_config.activation == "silu", "Only SiLU activation is supported."
+
+    workspace = marlin_make_workspace(hidden_states.device, max_blocks_per_sm=4)
```

Comment:
`marlin_make_workspace` has a `torch.zero()` operation, which requires to launch a cuda kernel. https://github.com/sgl-project/sglang/blob/9a426fc5ef9d8b93f8a12cd6425b27b90852f9f1/python/sglang/srt/layers/quantization/marlin_utils.py#L266

We can create a global static buffer as you did in `MarlinRunnerCore.run` to avoid it.